### PR TITLE
Propagate custom cfg directives to rustdoc and doctests

### LINKED ISSUE #1980: Cfgs from build script not passed to rustdoc
The `cfg` directives returned by the build script via the `cargo:rustc-cfg=foo` convention are not passed to `rustdoc` when invoking `cargo doc`.

`rustdoc` seems to accept them just fine i.e. invoking it manually with `--cfg foo` produces the expected results.

### PR DESCRIPTION
**Pull request by gkoz**
Fixes #1980

### CODE REVIEW

File: src/cargo/ops/cargo_rustc/mod.rs [hunk 1]
```diff
@@ -392,11 +398,20 @@ fn rustdoc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
...
     let name = unit.pkg.name().to_string();
-    let desc = rustdoc.to_string();
+    let build_state = cx.build_state.clone();
+    let current_id = unit.pkg.package_id().clone();
     let exec_engine = cx.exec_engine.clone();
 
     Ok(Work::new(move |desc_tx| {
-        desc_tx.send(desc).unwrap();
+        let outputs = build_state.outputs.lock().unwrap();
+        for (&(ref id, _), output) in outputs.iter() {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/mod.rs]: You'll want to do a hash lookup here to ensure that you're doing the right thing in a cross-compilation scenario. Right now this ignores the `Kind` which is an important distinction for target/host compiles.

The `Kind` to use in the key should be present in `unit.kind` right before the work is created, and that should do the trick!

File: src/cargo/ops/cargo_rustc/mod.rs [hunk 2]
```diff
@@ -392,11 +398,20 @@ fn rustdoc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
...
+            if *id == current_id {
+                for cfg in &output.cfgs {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/mod.rs]: Cargo currently prefers `for foo in bar.iter()` as opposed to `for foo in &bar`

**Comment by alexcrichton** [comment]
Awesome, thanks @gkoz!

File: src/cargo/ops/cargo_rustc/mod.rs (hunk 1 above)
Comment (gkoz) [inline src/cargo/ops/cargo_rustc/mod.rs]: Thanks, I wasn't sure whether the kind mattered here.

**Comment by gkoz** [comment]
Updated for both comments.

**Comment by gkoz** [comment]
@alexcrichton looks like you don't have `Rolling builds" enabled in appveyor project settings. Because of this it doesn't stop building stale commits promptly and takes longer to get to the current ones.

File: src/cargo/ops/cargo_rustc/mod.rs [hunk 3]
```diff
@@ -392,11 +398,18 @@ fn rustdoc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
...
 
     let name = unit.pkg.name().to_string();
-    let desc = rustdoc.to_string();
+    let build_state = cx.build_state.clone();
+    let key = (unit.pkg.package_id().clone(), unit.kind);
     let exec_engine = cx.exec_engine.clone();
 
     Ok(Work::new(move |desc_tx| {
-        desc_tx.send(desc).unwrap();
+        let outputs = build_state.outputs.lock().unwrap();
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/mod.rs]: This'll actually want to drop the lock ASAP instead of holding it over the entire call to `rustdoc`

Comment (gkoz) [inline src/cargo/ops/cargo_rustc/mod.rs]: Heh, right :flushed:

**Comment by alexcrichton** [comment]
Whoa, I had no idea that option existed! I've turned it on now

File: src/cargo/ops/cargo_rustc/mod.rs (hunk 3 above)
Comment (gkoz) [inline src/cargo/ops/cargo_rustc/mod.rs]: Would you put it like this then? It's a lot on one line and not very obvious that the lock will get released.
```rust
        if let Some(output) = build_state.outputs.lock().unwrap().get(&key) {
            for cfg in output.cfgs.iter() {
                rustdoc.arg("--cfg").arg(cfg);
            }
        }
```

Comment (gkoz) [inline src/cargo/ops/cargo_rustc/mod.rs]: But the alternative is putting everything into an explicit block, not very nice looking either.

Comment (gkoz) [inline src/cargo/ops/cargo_rustc/mod.rs]: Or
```rust
        if let Ok(outputs) = build_state.outputs.lock() {
            if let Some(output) = outputs.get(&key) {
                for cfg in output.cfgs.iter() {
                    rustdoc.arg("--cfg").arg(cfg);
                }
            }
        }
        /*
        else {
            return_an_error?
        }
        */
```

**Comment by gkoz** [comment]
I went with the more concise version giving the lock up after iterating over cfgs. At least I believe its lifetime will now be limited to the `if let` block.

**Comment by alexcrichton** [comment]
@bors: r+ 807da6b11e9a5821add1438247fbde3a6435dbe0

Thanks!

**Pull request merged**